A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,169 @@
+`ring_hash` policy unusable without an xDS infrastructure in place.
+
+This proposal extends the configuration of `ring_hash` policy to specify a
+metadata key. The value associated with this metadata key will be used as the
+request hash key if present. This will make it possible to use `ring_hash` by
+configuring it entirely in the [service config][service-config]. If this
+configuration is omitted, we will preserve the current behavior of using the xDS
+hash policy.
+
+### Using an explicit endpoint hash key
+
+Another limitation of the current `ring_hash` load balancing policy is that it
+always hashes the endpoint IP address to place the endpoints on the ring. In
+some scenario, this choice is not ideal: for example, [Kubernetes
+Statefulsets](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+offer a way to configure workloads with sticky identity such that endpoints keep
+storage and network identifier across restarts. However, the IP address may
+change across restarts. After a deployment, if all IP addresses have changed,
+then a completely new ring has to be constructed, even though it may have been
+desirable to keep the ring unchanged based on the Statefulsets identities, so
+that each instance stays at the same location on the ring.
+
+Envoy offers a solution to control endpoint hashes independently of IP
+addresses. This mechanism uses the `"envoy.lb"`
+[LbEndpoint.Metadata][LbEndpoint.Metadata] field `hash_key` value available in
+EDS instead of the endpoint IP address, as described in [the Envoy documentation
+for ring hash][envoy-ringhash].  This proposal adds support for setting the
+endpoint hash key explicitly via EDS by reusing the configuration mechanism
+implemented in Envoy. To retain the advantage of being able to use `ring_hash`
+without xDS, custom gRPC name resolvers will be able to set this endpoint
+attribute through the language specific resolver attribute interface.
+
+### Related Proposals:
+
+This proposal extends the following existing gRFC:
+
+* [gRFC A42: xDS Ring Hash LB Policy][A42]
+
+## Proposal
+
+### Explicitly setting the request hash key
+
+A new string field `request_metadata_key` will be added to the `ring_hash`
+policy config. The ring hash policy will be modified as follows:
+
+Upon loading the load balancing config, if the `request_metadata_key` field
+contains a value that is not a valid metadata key, then the configuration is
+rejected. If the `request_metadata_key` refers to a binary metadata (suffixed
+with `-bin`), the configuration is also rejected.
+
+At pick time:
+- If `request_metadata_key` is not empty, and the request metadata has a
+non-empty value, then the request hash key will be set to this value. If
+`request_metadata_key` contains multiple values, then values are joined with a
+comma `,` character between each value before hashing.
+- If `request_metadata_key` is not empty, and the request has no value or an
```

Comment:
We finally had a chance to chat about this today.  In addition to the two options I mentioned above, we came up with a few more options that would all be implemented purely within the ring_hash picker:

3. For a request that does not have the specified header, we compute a random hash in the picker.  If we trigger a connection attempt using the normal ring_hash picker logic, then before we queue the pick, we scan forward through the entire ring searching for a subchannel in state READY.  If we find one, we return it; otherwise, we queue the pick.  If we wind up queuing the pick, then we will still compute a new random hash on the next pick attempt and therefore trigger another connection attempt, but if there is at least one subchannel in state READY, then we won't need a second pick for the RPC, which means that we'll have eliminated the problem of increasing RPC latency.  And while this doesn't completely eliminate the problem of creating unnecessary connections, it should significantly reduce the number in the common case where connection attempts finish quickly.  (Also, note that there are already some other cases in ring_hash where we generate unnecessary connections, so this isn't a purely new problem.)  This option would result in slightly worse load distribution for these RPCs while there are few existing connections, but it converges to fairly reasonable behavior with a reasonable rate of traffic.
4. We could just fail the pick for an RPC without the specified header.  This seems sub-optimal, since we'd generally prefer the lack of a header to affect load balancing but not correctness.  (Also, note that for wait_for_ready RPCs, this would result in the RPC remaining queued until its deadline, unless the picker generates a "drop" result instead of a "fail" result.)
5. We could treat a missing header as being present but having the empty string for the value.  This would result in all RPCs without the header being hashed to the same endpoint, which is a significant difference to the current behavior of choosing randomly.  (Another way to say this is that ring_hash would basically devolve to pick_first for this case.)

Given this menu of options, we think that option 3 is the best trade-off in terms of how much work it is to design and implement relative to how close it gets us to ideal behavior.

It's also worth noting that all three of these new options are purely implemented within the ring_hash picker, which means that we can pick any one of them now, and then later we could add a config option to configure one of the other two options instead, if that turns out to be necessary.

What do you think?

Reply:
Thank you for providing alternatives. Reading through option 3, I realize that you are concerned with usages of ring hash where a given channel connects to a subset of the endpoints on the ring. This is not something I had really considered: the primary use case we have is one where a given channel ends up connecting to all endpoints anyway (just like round robin), and ring hash provides affinity between the content of a header and the endpoint that gets picked.

Option 3 sounds ideal in terms of "theoretical" properties (converging to random load balancing when many requests lack the header, without a single request causing multiple new connections), but it also sounds a bit complicated to me. I find it a bit hard to wrap my head around edge cases, and I don't look forward to having to explain it to our users. I don't have a good sense of whether its behavior will need to be understood and troubleshooted.

I like the simplicity of option 4 actually. At least for our use case, which is to afford some locality based on request content, if the application fails to provide the header, the fall back to a random endpoint is likely to overload the system: the system may not be designed to resist lacking any locality. If users want the behavior of using a random endpoint when the request is lacking the header, then they can set it to a random value in the application.

I would rule out option 5 because if the traffic suddenly starts lacking the header, then all requests from all instances of the channel will end up picking the endpoint closest to `xxhash("")` on the ring, and there would be a high risk of overloading it, then the next closest, etc. This seems really dangerous, but maybe I'm missing something.

If you're all not too concerned about the complexity of option 3, then I'm happy to go with it, otherwise I would go for option 4.

Reply:
OK, that all makes sense. I updated the proposal to use option 3.